Add dataframe mixin

choldgraf opened the pull request:
This is an incomplete PR as there are still changes to be made, but I wanted to put this up there to see if this is what you guys had in mind. I've added a DataFrameMixin to the base module, and then subclassed Epochs to this. I also added in a few checks to decide what indices to give the multi-index object, but this is probably too Epochs-specific right now (e.g., it's currently calling shape[2] in several places even though this won't exist if we're using a Raw instance). Let me know how this looks / any suggestions, and I'll make changes accordingly.

dengemann commented on mne/io/base.py:
```diff
@@ -1822,3 +1823,108 @@ def _check_update_montage(info, montage):
+        if isinstance(self, (Epochs, EpochsArray)):
+            default_index = ['condition', 'epoch', 'time']
+        else:
+            default_index = ['condition', 'time']
```
to be handled in addition: Raw, Evoked, SourceEstimate

dengemann commented on mne/io/base.py:
```diff
@@ -1822,3 +1823,108 @@ def _check_update_montage(info, montage):
+        data = self.get_data()[:, picks, :]
+        shape = data.shape
+        data = np.hstack(data).T
+        if copy:
```
we follow the convention to make explicit the conditional and to expect bool types: `if copy is True`

choldgraf commented:
OK, some more changes made, let me know what you think. Do you think that any new tests need to be made, other than re-naming tests to use "to_data_frame" rather than "as_data_frame"?

Also, I haven't tested out the SourceEstimate case yet, because I don't really know what that is and couldn't figure out how to construct one with dummy data (I kept getting a vertices ordering error). I can try to work this out if someone points me in the right direction, but have assumed that it behaves similarly to Evoked classes.

As an aside, in testing this out I noticed that the EpochsArray allows you to construct an object with different numbers of events in the data array vs. the events id array. E.g., you can do something like:

```
n_events = 5
data = np.random.randn(n_events, 10, 1000)
info = mne.create_info(['{0}'.format(i) for i in range(10)], 100,
                       ch_types=['eeg' for _ in range(10)])
events = np.array([[10, 0, 2],
                   [20, 0, 1]])
epochs = mne.EpochsArray(data, info, events)
```
I feel like this should throw an error, but maybe I'm not noticing something?

Eric89GXL commented:
I think that should throw a warning -- there might be cases where users have lots of event types, only some of which exist for each subject, as in the case of press-dependent event numbering.

Here is a usable SourceEstimate of random data for the subject `fsaverage`:
```
stc = mne.SourceEstimate(np.random.randn(20484, 5),
                         [np.arange(10242), np.arange(10242)],
                         0, 0.001, 'fsaverage')
```

Eric89GXL commented:
If you can add simple tests for all cases, that would be good. It could just be exporting the data, and making sure the right data shape and indices came through.

choldgraf commented:
Will do - have some work to catch up on today, but I have a flight tomorrow
morning so I'm hoping to work out some of these PRs then. That said, the
flight is to New Orleans for a bachelor party, so I will be very
unproductive after I land :)

On Mon, Jan 26, 2015 at 10:04 AM, Eric Larson <notifications@github.com>
wrote:

> If you can add simple tests for all cases, that would be good. It could
> just be exporting the data, and making sure the right data shape and
> indices came through.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1760#issuecomment-71506790>.
>

dengemann commented:
2015-01-27 18:46 GMT+01:00 Chris Holdgraf <notifications@github.com>:

> Will do - have some work to catch up on today, but I have a flight
> tomorrow
> morning so I'm hoping to work out some of these PRs then. That said, the
> flight is to New Orleans for a bachelor party, so I will be very
> unproductive after I land :)
>

:)


>

On Mon, Jan 26, 2015 at 10:04 AM, Eric Larson <notifications@github.com>
> wrote:
>
> > If you can add simple tests for all cases, that would be good. It could
> > just be exporting the data, and making sure the right data shape and
> > indices came through.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/mne-tools/mne-python/pull/1760#issuecomment-71506790>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1760#issuecomment-71692739>.
>

agramfort commented:
make the bachelor party a hackathon it will make it really unique and
memorable :)

dengemann commented:
Where are we here? Any chance we can merge this soonish? Let me know if you need help.

dengemann commented:
#1837

choldgraf commented:
Hey - let me take a look at this and see what's left to be done. I'll get
back in the next day or two.

On Sun, Mar 8, 2015 at 7:23 PM, Denis A. Engemann <notifications@github.com>
wrote:

> Where are we here? Any chance we can merge this soonish? Let me know if
> you need help.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1760#issuecomment-77790999>.
>

choldgraf commented:
So as far as I can tell, the main thing would be to add tests for this. As a first step I tried modifying the test_epochs.py function so that it uses to_data_frame instead of as_data_frame. As you mentioned before, all functionality should be pretty much the same thing otherwise.

A single line of code fails (if I comment it out, then it works):

    assert_raises(ValueError, epochs.to_data_frame, np.arange(400))

However, I'm not sure why. I get the following error which makes me think it's not due to my change in code:

    File "/Users/choldgraf/github/mne-python/mne/io/base.py", line 102, in to_data_frame
    if not in1d(picks, np.arange(len(self.events))).all():
    NameError: global name 'in1d' is not defined

I've updated the PR to reflect this change in tests.

Other than this, it seems like what's left to do is:

1. Import the DataFrameMixin to these objects: TFRAverage, SourceEstimate
2. Make sure that the dataframe tests that were already written for these now use to_dataframe

Does this sound right to you? Also, as for changing the name from "as_data_frame" to "to_data_frame", should I just remove the old function, or point it to the new one?

dengemann commented:
@choldgraf silly question, is in1d imported from mne.fixes?

dengemann commented:
> Does this sound right to you? Also, as for changing the name from "as_data_frame" to "to_data_frame", should I just remove the old function, or point it to the new one?

We should deprecate it. say `git grep derpecated` in the command line to get an idea of the pattern.

choldgraf commented:
@choldgraf silly question, is in1d imported from mne.fixes?

Ah, yes! I didn't add the import to base as well. I should either take a break or drinks some more tea :)

choldgraf commented:
Is the merge failing because I've renamed the test?

dengemann commented:
@choldgraf probabably because the MNE master branch is already in the future, you want to insert the past into its history. You need to switch to master, pull, switch back to your branch and rebase master to get your commits on top, such that they represent the future as seen from the master branch.

choldgraf commented:
ok, definitely going to need a cup of tea for that. I'll tell you how it goes.

dengemann commented:
@choldgraf the reason why this matters is that epochs.py has changed on master. Git does not know how to deal with it. Prepare a good cup of tea and ping us if you need some contact. 

The commands you need are

git checkout master
git pull upstream master
git checkout add_dataframe_mixin
git rebase master

cc https://help.github.com/articles/resolving-merge-conflicts/

choldgraf commented:
Ah I see. When I tried to rebase I got an error in merging epochs.py Need to figure out how to inspect exactly what went wrong, but I'm working on it.

dengemann commented:
@choldgraf this may seem scary, but it's simple.

Spot the lines that are no Python code.
In the first line you would find `HEAD`, followed by the code diff on master. then a separator ====.
then your conflicting code. then the commit hash.

You need to remove all 3 lines that are no valid Python code and be diplomatic. Prefer the code that belongs to HEAD unless you wanted to change it. In the latter case make your intended modifications. Once you're done pay a visit to git and add the file you just edited, for example:

git add mne/epochs.py

then tell git to `git rebase --continue`

and repeat this procedure if needed.

Eric89GXL commented:
I prefer using `git mergetool` to take care of that stuff (removing non-python lines, actually merging the files, and doing the `git add`) for me, so my iteration is usually
```
git fetch upstream
git rebase upstream/master
git mergetool
git rebase --continue
git mergetool 
git rebase --continue
...
```

Eric89GXL commented:
`meld` is a particularly useful tool for such 3-way merges

dengemann commented:
Did not know about mergetool + meld, thanks @Eric89GXL

choldgraf commented:
Cool, I think it worked. Didn't think about using a mergetool but it makes sense since rebasing is basically just doing a ton of merges in a row, right? The conflict was in the imports, but should be fixed now.

dengemann commented:
@choldgraf yes looks clean!

dengemann commented:
@choldgraf once Travis is happy, can you build docs or run the related examples manually to see if we're good?

choldgraf commented:
Well, the test that I've got is just for epochs, I should add/modify the respective tests for Evoked, Raw, SourceEstimate, and TFRAverage, no?

dengemann commented:
>  I should add/modify the respective tests for Evoked, Raw, SourceEstimate, and TFRAverage, no?

at least minimally to make sure all conditionals inside the method get executed and return.

choldgraf commented:
ok, I'll add the mixin for SourceEstimate and TFRAverage, will update
tests, and will try to run examples tomorrow. (or tonight if I've got some
extra time after getting my other work done)

On Tue, Mar 10, 2015 at 6:44 PM, Denis A. Engemann <notifications@github.com
> wrote:

> I should add/modify the respective tests for Evoked, Raw, SourceEstimate,
> and TFRAverage, no?
>
> at least minimally to make sure all conditionals inside the method get
> executed and return.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1760#issuecomment-78185332>.
>

dengemann commented:
> ok, I'll add the mixin for SourceEstimate and TFRAverage, will update
tests, and will try to run examples tomorrow. (or tonight if I've got some
extra time after getting my other work done)

brilliant! Looks like we're very close.

agramfort commented on mne/io/base.py:
```diff
@@ -42,8 +44,135 @@
+            Scaling to be applied to the channels picked. If None, defaults to
+            ``scalings=dict(eeg=1e6, grad=1e13, mag=1e15, misc=1.0)`.
+        copy : bool
+            If true, data will be copied. Else data may be modified in place.
```
docstring is not up to date

agramfort commented:
I am happy if travis and @dengemann are happy

choldgraf commented:
Question - I'm looking into the SourceEstimate code now, and I wonder if this is worth having its own to_data_frame method. The reason for this is that SourceEstimate seems to be unique in that columns are vertices, as opposed to "time". The ToDataFrameMixin is already becoming a bit clunky since there are lots of checks (e.g., if it's Epochs, do this, if it's Raw, do that). Would we rather have everything in one mixin, or keep the SourceEstimate separate?

choldgraf commented:
One other thing - it looks like the AverageTFR doesn't currently have an "as_data_frame" method. For the sake of getting this PR done, should that extra functionality come with another PR?

dengemann commented:
@choldgraf you can handle all this in the mixin.

Special operations go in conditionals in which you test if self is instance of class SourceEstimate, etc.

But you can add the support for AverageTFR later.

choldgraf commented:
OK, tests are passing for epochs and evoked. The Raw tests seem to skip, maybe this is because loading data is required? How can I make this continue? After this I think I just need to add the SourceEstimate and corresponding tests.

Eric89GXL commented:
Try mne.datasets.testing.data_path(), it should trigger testing data
download

agramfort commented:
or use the tiny raw fix file we ship in the repo

Eric89GXL commented on mne/io/base.py:
```diff
@@ -33,17 +33,156 @@
+    '''Class to add to_data_frame capabilities to certain classes.'''
+    def to_data_frame(self, picks=None, index=None, scale_time=1e3,
+                      scalings=None, copy=True, start=None, stop=None):
+        """Get the epochs as Pandas DataFrame
```
s/epochs/data

Eric89GXL commented on mne/io/base.py:
```diff
@@ -33,17 +33,156 @@
+        if isinstance(self, (Epochs, EpochsArray)):
+            default_index = ['condition', 'epoch', 'time']
+        elif isinstance(self, (RawFIFF, RawArray, Evoked,
+                               EvokedArray, SourceEstimate)):
```
I think you want `_BaseRaw` instead of enumerating the raw classes, same with `_BaseEpochs` instead of enumerating epochs types

Eric89GXL replied:
and instead of doing multiple checks against types in multiple places, can you do them all here, once, and do all the necessary triaging so that the rest of the code can run without instance-style triaging?

choldgraf replied:
I was trying to think about how to do this without lots of type-switches in there, but was having a hard time figuring out how to do this effectively. E.g., in all cases, you need to define "data" and "times". This is different for the different types. However, in order to get "data" you need picks, and the catch to make sure picks is properly created depends on data type as well. E.g., 

        if isinstance(self, _BaseEpochs):
            default_index = ['condition', 'epoch', 'time']
            picks_check = self.events
        elif isinstance(self, (RawFIFF, RawArray, Evoked, SourceEstimate)):
            default_index = ['time']
            picks_check = self.ch_names

        if index is not None:
            _check_pandas_index_arguments(index, default_index)
        else:
            index = default_index

        if picks is None:
            picks = list(range(self.info['nchan']))
        else:
            if not in1d(picks, np.arange(len(picks_check))).all():
                raise ValueError('At least one picked channel is not present '
                                 'in this object instance.')

        if isinstance(self, _BaseEpochs):
            data = self.get_data()[:, picks, :]
            times = self.times
            n_epochs, n_picks, n_times = data.shape
            data = np.hstack(data).T  # (time*epochs) x signals

It'd be good if I could move the `isinstance` for `_BaseEpochs` up top to the switch, but it needs picks to be defined first, and that's something that's shared across all object types. Does that make sense?

dengemann replied:
I'm not sure I can follow ...

Eric89GXL replied:
so move the `picks` definition before the triaging...?

In a worst case, I think I'd rather see one or two lines replicated within the checks than see the checks replicated several times. Neither one will be DRY but the checks only being done once is nice because then the rest of the code is clearer. You could even export the "same across all objects" bit to a small helper function, which would make it pretty DRY. I guess I have a slight preference for "get the information from each type of object that we need, then follow the same code path for export" over "get information out of these objects, do some stuff, get other information out of these objects, do some stuff, get more information out of these objects, do some stuff, export complete". The latter is what the code was starting to look like to me, so it was harder to read than what I imagine the former will look like.

Eric89GXL replied:
...and adding a case for export is also much easier if the type-specific code is all contained in one place.

Eric89GXL replied:
Just read your edit -- I think a `picks = _get_check_picks(self)` helper call within each `if` would work well here.

FYI GitHub does not send notices about edits, so they are easy to miss if you rely on emails to track conversations, as many people do. For this reason I try to only use edits only for things I don't mind if people don't ever see (e.g. to correct typos).

choldgraf replied:
Good to know re: comments. I'll just add another line note from now on.

So you imagine a helper function _get_check_picks(self) that has a switch in it which returns the correct picks type based on the type of self? That feels like cheating to me :)

Eric89GXL replied:
not quite. The idea is that the picking is independent of type, right? If so I mean something like this:
```
    def _get_check_picks(self, picks, picks_check):
        """Helper to check picks"""
        if picks is None:
            picks = list(range(self.info['nchan']))
        else:
            if not in1d(picks, np.arange(len(picks_check))).all():
                raise ValueError('At least one picked channel is not present '
                                 'in this object instance.')
        return picks

...

        # triage to extract data (using picks), times, and default_index
        if isinstance(self, _BaseEpochs):
            default_index = ['condition', 'epoch', 'time']
            picks = self._get_check_picks(picks, self.events)
            data = self.get_data()[picks, :, :]
            times = self.times
            n_epochs, n_picks, n_times = data.shape
            data = np.hstack(data).T  # (time*epochs) x signals
        elif isinstance(self, (_BaseRaw, Evoked)):
            default_index = ['time']
            picks = self._get_check_picks(picks, self.ch_names)
            if isinstance(self, _BaseRaw):
                data, times = self[picks, start:stop]
            else:  # isinstance(self, (Evoked, SourceEstimate)):
                data = self.data[picks, :]
            times = self.times

        # Now our code is unified (hooray!)
        n_picks, n_times = data.shape
        data = data.T
        ...
```
Or something along these lines. I'm not 100% sure what all these lines were meant to do or why, but hopefully you get the idea of where I'm heading with it.

Eric89GXL replied:
Note that everything after the `...` in the code block is meant to be thought of as living in the `to_data_frame` function.

Eric89GXL commented on mne/io/base.py:
```diff
@@ -33,17 +33,156 @@
+            picks = list(range(self.info['nchan']))
+        else:
+            if isinstance(self, (Epochs, EpochsArray)):
+                if not in1d(picks, np.arange(len(self.events))).all():
```
I don't think this line is correct anyway (which is why I was confused during re-coding). `picks` has to do with the number of channels, `self.events` has to do with the number of epochs... I'm not sure why you'd compare them.

Eric89GXL commented on mne/io/base.py:
```diff
@@ -33,17 +33,156 @@
+                                     'in this Evoked instance.')
+
+        if isinstance(self, (Epochs, EpochsArray)):
+            data = self.get_data()[:, picks, :]
```
and later you use it to index into the channels dim

Eric89GXL commented on mne/io/base.py:
```diff
@@ -33,17 +33,156 @@
+            names = [id_swapped[k] for k in self.events[:, 2]]
+            mindex.append(('condition', np.repeat(names, n_times)))
+            mindex.append(('epoch', np.repeat(np.arange(n_epochs),
+                          n_times)))
```
And again, `mindex` could be constructed above in the already-done triaging

Eric89GXL commented on mne/io/base.py:
```diff
@@ -33,17 +33,156 @@
+        df = pd.DataFrame(data, columns=col_names)
+        [df.insert(i, k, v) for i, (k, v) in enumerate(mindex)]
+        if index is not None:
+            with warnings.catch_warnings(record=True):
```
Please add a comment about why this is necessary. It's better to avoid warnings altogether, so it's important to justify why they are being passed silently when it's done.

Eric89GXL commented on mne/io/base.py:
```diff
@@ -33,17 +33,156 @@
+                if 'time' in index:
+                    df['time'] = df['time'].astype(np.int64)
+                df.set_index(index, inplace=True)
+        if index is default_index:
```
`is` is a pretty strong comparison -- what does this conditional do? Could it be more clearly be expressed some other way?

choldgraf replied:
I think that most of these bits of code are not things that I inserted, but just copied them over when compiling all the `as_data_frame` methods into this one. I will take a pass to try and clean this up but it sounds like it may take me a little while to figure out.

Eric89GXL replied:
You can try using `git blame` on `master` to see who most recently touched the original lines in question. It's also possible the version in `master` is doing something wrong / the checks aren't quite right.

choldgraf replied:
Assuming that I'm doing the git blaming correctly, it looks like @dengemann along with someone named Nick Ward was involved in the warning code, and @dengemann wrote the original `if not in1d(picks, np.arange(len(self.events))).all():` code back in 2012, though it looks like you and someone named t3on also edited this section a bit in 2015 (I think it was just to change the warning message maybe?)

Eric89GXL replied:
Yeah if I touched the code it's probably just to clean it up. @dengemann any idea why `picks` is tested against `self.events` for `Epochs`? It seems like it should be tested against the channels, not the events.

choldgraf commented:
OK, I made some triaging edits, picks edits, and warnings catching edits per comments from @Eric89GXL. Here's a quick breakdown:

1. Add picks check to the top of the triage because I'm using `self.ch_names` in every case now. 
2. Removed the warning and added a print statement that time column is being converted to int64 (not sure how you guys feel about print statements but at least this is more verbose?)
3. Triaged so that all of the object-specific stuff is done with a single if/then section

Still need to add SourceEstimate, but at least the current code works on test_evoked and epochs

choldgraf commented:
@agramfort could you give a bit more direction on using the "tiny" data set? I tried downloading the full dataset with `data_path()` but running nosetests keeps skipping the function. I did some searches for "tiny" data set in the repo but couldn't find anything obvious...

agramfort commented:
see

https://github.com/mne-tools/mne-python/blob/master/mne/io/tests/data/test_raw.fif

HTH

choldgraf commented:
TY - I'll get this working soon

On Sun, Mar 15, 2015 at 2:56 PM, Alexandre Gramfort <
notifications@github.com> wrote:

> see
>
>
> https://github.com/mne-tools/mne-python/blob/master/mne/io/tests/data/test_raw.fif
>
> HTH
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1760#issuecomment-81257246>.
>

choldgraf commented:
OK, I think SourceEstimate is working now. This object has a pretty different structure from the Epochs/Evoked/Raw objects, so I had to add another switch to do special stuff with this. While the code passes tests, IMO the ToDataFrameMixin code is a little bit "indenty" right now. Maybe there's a smarter way of doing this than I've done here? Either way, thoughts are appreciated!

Eric89GXL commented:
It looks like you've done just about all you can to unify the arguments / simplify the conditionals. Sometimes it takes a lot of indentations, but to me the flow is still pretty clear. Even without you saying so, it's clear that SourceEstimates are different from Evoked/Epochs (which makes sense because one is source vs sensor space), then within evoked/epochs there are other differences -- and they are unified as soon as possible to continue onward. Nice!

Eric89GXL commented:
Ping me once the CIs are happy and I'll merge. I imagine this will hurt our coverage a bit because the deprecated functions aren't checked, but that's okay because we can quickly remove them post-release.

Eric89GXL commented:
Can you just check that test time isn't adversely affected? You can do e.g.
```
nosetests mne/tests/test_source_estimate --with-timer
```
on `master` and then on your branch and make sure the tests you modified didn't increase in time. They shouldn't have, since you're doing very similar operations.

choldgraf commented:
Here are the results for three runs of each on source_estimate:

On master: .296, .284, .281
This branch: .299, .288, .281

And for test_epochs (now only testing test_data_frame because the whole file takes a little while to run):
On master: 2.30, 2.08, 2.07
On this branch: 2.54, 2.55, 2.53

How much would be too much of a difference?

Eric89GXL commented:
Every little bit counts, but that doesn't look too bad to me.

choldgraf commented:
So looking at the appveyor build, it seems like the docstring is messed up. I think this is because I've got both start and stop on the same line in the docstring for to_data_frame...does that sound right?

e.g., from the error logs:
`mne.io.base.Epochs.to_data_frame in doc ['start, stop']`

Eric89GXL commented:
Yeah, that should be split

choldgraf commented:
OK, I split up the docstring. Not sure what's up with the failing tests...it looks like with travis, the build that fails is skipping the to_data_farme tests because they require pandas, the other build is failing with error `urllib2.URLError: <urlopen error [Errno 11001] getaddrinfo failed>`. It seems like this isn't related to any of these changes but correct me if I'm wrong?

Eric89GXL commented:
Any test that requires `pandas` needs the `requires_pandas` decorator to be skipped on builds and user systems that don't have it.

Don't worry about the getaddrinfo failure, that's not due to your changes.

Eric89GXL commented:
Oh wait you have those decorators, your real problems are style-related:

https://travis-ci.org/mne-tools/mne-python/jobs/54632990

Eric89GXL commented:
@choldgraf is this ready to go from your end? If so, I'll merge by rebase.

choldgraf commented:
If tests are passing then I am happy to merge!

dengemann commented on mne/io/base.py:
```diff
@@ -33,17 +33,173 @@
+        if index is not None:
+            if 'time' in index:
+                print('Converting time column to int64...')
+                df['time'] = df['time'].astype(np.int64)
```
@choldgraf what's this? please use the logger instead.

choldgraf replied:
It use to be logged and @Eric89GXL had some concerns about this. I am happy to change back...

dengemann replied:
@Eric89GXL why the print?

Eric89GXL replied:
I don't remember asking for a `print`, in any case I'll change it to `logger.info`

choldgraf replied:
you didn't :)

dengemann commented on mne/io/base.py:
```diff
@@ -33,17 +33,173 @@
+    def to_data_frame(self, picks=None, index=None, scale_time=1e3,
+                      scalings=None, copy=True, start=None, stop=None):
+        """Export data in tabular structure as a pandas DataFrame.
+        Columns and indices will depend on the object being converted.
```
please insert one empty line between the header and the body

Eric89GXL commented:
@dengemann let me know when you're done commenting, @choldgraf I'll take care of the last few points assuming they're minor (I already rebased your branch)

choldgraf commented:
OK, please let me know if there's anything else I can do. Thanks to @dengemann @Eric89GXL and @agramfort for being patient with me on this first pull request of mine :)

dengemann commented:
@Eric89GXL I think I'm fine. @choldgraf Thanks for your efforts and patience with us!

Eric89GXL commented:
Merged by rebase, thanks @choldgraf! I also added you to the whats_new.py
